### PR TITLE
Implement `cupy.flatiter`

### PR DESCRIPTION
**Pull request by takagi**
Follows #3139 .

I will post several PRs to implement `cupy.flatiter`. It will have similar interface to `cupy.ndarray` as `numpy.flatiter` does to `numpy.ndarray`. For the first, this PR implements its subscript operators so that we can replace `cupy.fill_diagonal`’s implementation using it.

I suppose that `cupy.flatiter` would not provide its iteration interface, as for-loop or the `next` method,  in contrast to `numpy.fatiter` because iteration is essentially serial concept and does not suit to GPU’s pararell computation model.

### CODE REVIEW

**Comment by takagi** [comment]
@emcastillo Please let me ask your advice about `cupy.flatiter`'s iteration interface later.

**Comment by emcastillo** [comment]
As long as the iterator returns `cupy.ndarray` objects I think it makes sense.
Returning the actual values will synchronize the device in every iteration and lose flexibility

**Comment by takagi** [comment]
Sure, I will work as such in another PR!

File: cupy/indexing/iterate.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                size = size // s
+                indices.append(ind // size)
+                ind %= size
+            base[tuple(indices)] = value
+            return
+
+        if isinstance(ind, slice):
+            base = self._base
+            s = internal.complete_slice(ind, base.size)
+            s_start = s.start
```
Comment (emcastillo) [inline cupy/indexing/iterate.py]: nit: use `size = s.stop - s.start` instead of tracking both

Comment (takagi) [reply cupy/indexing/iterate.py]: Fixed!

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**